src/query/storage/m3/config.go

```diff
@@ -313,3 +316,77 @@ func (c ClustersStaticConfiguration) NewClusters(
 	return NewClusters(unaggregatedClusterNamespace,
 		aggregatedClusterNamespaces...)
 }
+
+// NewDynamicClusters instantiates a new Clusters instance that pulls
+// cluster information from etcd.
+func (c ClustersStaticConfiguration) NewDynamicClusters(
+	instrumentOpts instrument.Options,
+	opts ClustersStaticConfigurationOptions,
+) (Clusters, error) {
+	clients := make([]client.Client, 0, len(c))
+	for _, clusterCfg := range c {
+		clnt, err := clusterCfg.newClient(client.ConfigurationParameters{
+			InstrumentOptions: instrumentOpts,
+		}, opts.CustomAdminOptions...)
+		if err != nil {
+			return nil, err
+		}
+		clients = append(clients, clnt)
+	}
+
+	// Connect to all clusters in parallel
+	var (
+		wg   sync.WaitGroup
+		cfgs = make([]DynamicClusterNamespaceConfiguration, len(clients))
+
+		errLock  sync.Mutex
+		multiErr xerrors.MultiError
+	)
+	for i, clnt := range clients {
+		i := i
+		clnt := clnt
+		nsInit := clnt.Options().NamespaceInitializer()
+
+		// TODO(nate): move this validation to client.Options once static configuration of namespaces
+		// is no longer allowed.
+		if nsInit == nil {
+			return nil, errNoNamespaceInitializerSet
+		}
+
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			cfg := cfgs[i]
+			cfg.nsInitializer = nsInit
```

Comment:
Hm I'm not sure that this will necessarily update this? May need to explicitly `cfgs[i] = cfg` at the end of this functionm and add a test to ensure these are set correctly?

Reply:
Yeah, I'll fix this up. I was just going to add a test once I do the integration, but I think I'll just create test implementation of Clusters we can use here to put this logic through it's paces.